### PR TITLE
E2E: Undo/Redo tests

### PR DESCRIPTION
**Pull request by 4ydan**
Testing Undo/Redo with opencv images of the canvas.

Rebased on top of planting e2e tests #766 
I added mypy, flake8 and black commit hooks.

## Basics

- [x] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildserver is happy.

## Checklist

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed the introduction tour
- [ ] I wrote migrations in a way that they are compatible with already present data
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Comment by markus2330** [comment]
Isn't it possible to click on the canvas and check text in the left toolbox? How exactly plants looks like easily changes in the future.

**Comment by 4ydan** [comment]
That was also my first idea, but the issue is after pressing undo, the select thingy is still there and clicking again on the same pixel doesnt unselect and remove the plants name from the left bar.

image

**Comment by markus2330** [comment]
Yes, I agree. Let us fix this first and then continue with this E2E test.

Can you report the issue? If it is a small fix, it would be great to have it in the release.

**Comment by 4ydan** [comment]
Done #772

**Comment by markus2330** [comment]
Thx!

**Comment by 4ydan** [comment]
I rebased this on top of #784, for smoother merging I think?

**Comment by markus2330** [comment]
> I rebased this on top of https://github.com/ElektraInitiative/PermaplanT/pull/784, for smoother merging I think?

Yes, always feel free to pull in PRs you need. In this case, I want to merge #784 soon anyway, as it is relevant for version 0.3.2.

File: e2e/steps/test_undo_redo.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+
+@when("I plant something")
+def plant_something(mpp: MapPlantingPage):
+    plant_a_tomato(mpp)
+
+
+@when("I press undo")
+def press_undo(mpp: MapPlantingPage):
+    mpp.click_undo()
```
Comment (markus2330) [inline e2e/steps/test_undo_redo.py:34]: Try to avoid to use different terms for the same thing. This seems to be the case for click/press.

File: e2e/steps/test_undo_redo.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    mpp.click_on_canvas()
```
Comment (markus2330) [inline e2e/steps/test_undo_redo.py:35]: Doesn't this belong to the step "my plant is gone"?

File: e2e/steps/test_undo_redo.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+@when("I accidentally pressed undo after planting one plant")
+def accidental_undo(mpp: MapPlantingPage):
+    plant_a_tomato(mpp)
+    mpp.click_undo()
+
+
+@when("I press redo to get my plant back")
+def press_redo(mpp: MapPlantingPage):
+    mpp.click_redo()
+    mpp.click_on_canvas()
```
Comment (markus2330) [inline e2e/steps/test_undo_redo.py:55]: same here.

**Review by markus2330 — APPROVED** [review]
LGTM but needs rebase

**Comment by markus2330** [comment]
Please always feel free to readd "please merge" once you are ready again.

File: e2e/steps/test_undo_redo.py (hunk 1 above)
Comment (4ydan) [reply e2e/steps/test_undo_redo.py:34]: True, thank you, I keep doing that.

File: e2e/steps/test_undo_redo.py (hunk 2 above)
Comment (4ydan) [reply e2e/steps/test_undo_redo.py:35]: Good point, ty.

File: e2e/conftest.py [hunk 4]
```diff
@@ -8,6 +9,15 @@
 from e2e.pages.maps.edit import MapEditPage
 from e2e.pages.maps.planting import MapPlantingPage
 
+load_dotenv()
+
+E2E_TIMEOUT: float = float(os.getenv("E2E_TIMEOUT", 30000))
+E2E_WIDTH: int = int(os.getenv("E2E_WIDTH", 1920))
+E2E_HEIGHT: int = int(os.getenv("E2E_HEIGHT", 1080))
```
Comment (markus2330) [inline e2e/conftest.py:16]: Please try to keep as simple as possible. If there is no need to have e.g. a setting differently locally and in deployment, we also don't need to have it configurable.

Comment (4ydan) [reply e2e/conftest.py:16]: Alright.

**Comment by 4ydan** [comment]
Tests seem to fail in Jenkins but pass remotely, I increased the timeouts. Hopefully this helps.

**Comment by 4ydan** [comment]
To sum up what has been done here additionally to the new tests.

- new pytest retry plugin (bigger community, better options)
  `--reruns 2 --reruns-delay 5 --rerun-except AssertionError`
- reduced plant search tests scenarios
A mix between partial and exact searches, should give us the best of both worlds and reduce test time a bit.
- a 30seconds timeout on all web actions should make tests overall more robust

**Comment by 4ydan** [comment]
Ah i see, because of `--rerun-except AssertionError` tests that fail with assertions dont get retried.
Since checking the pages title is an assertion, these tests will not see any retries.
A very popular example is the login procedure, which seemed to fail a lot in the past but passed with enough retries, maybe its time to look into this.

**Comment by 4ydan** [comment]
Is it possible that api/config is a bad check for backend activity, since its not using the database connection?

**Comment by 4ydan** [comment]
Unluckily the front page gets hard reloaded after login. Maybe all the image traffic is an issue for the tests.

**Comment by 4ydan** [comment]
Jenkins build please

**Comment by 4ydan** [comment]
Image loading seems also to be an issue when planting something.

**Comment by markus2330** [comment]
> Is it possible that api/config is a bad check for backend activity?

Mmmh, it doesn't use the database so it could be that the database is not yet initialized, even though the database initialization is one of the first things done by the backend.

> Or is it maybe too much load on the backend?

Mmmh, the backend is stateless and the timeout is 30 sec.

I think both things are rather unlikely. If they were a problem, it would be a bug in the backend.

I see following error in the backend logs:

```
[2023-08-06T10:48:09Z ERROR backend::error] Error executing diesel SQL query: duplicate key value violates unique constraint "map_unique_name"
[2023-08-06T10:48:09Z INFO  actix_web::middleware::logger] 127.0.0.1 "POST /api/maps HTTP/1.1" 500 64 "https://pr.permaplant.net/maps/create" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/115.0.5790.75 Safari/537.36" 0.000891
```

So more likely is that two test cases use the same map name and they are executed in parallel.

**Comment by 4ydan** [comment]
I'm pretty confident I found part of the problem. It seems like the image loading and my unexperience with frontend programming (page loading, doms, events, etc) introduced some problems with bad timings (test clicks on plant, but image of the plant is not loaded yet).

The solution I introduced for now is this:

```
def dont_load_images(self):
        """
        Ignore all png/jpg images on all pages.
        """
        self.page.route(
            re.compile(r"\.(jpg|png)$"),
            lambda route: route.fulfill(status=404, content_type='text/plain', body='Not Found!')
        )
```

Its a bit strict, but it does the job for now.

> I see following error in the backend logs:

This is only because when a test starts a retry, he creates the same map with the same name again.
I tried solving this in the past, but its too much effort marking maps with timestamps or something.

**Comment by 4ydan** [comment]
So these two methods are pretty good
```
self.page.set_default_timeout(timeout=E2E_TIMEOUT)
expect.set_options(timeout=E2E_TIMEOUT)
```
First I did not know about `expect.set_options()` and I was confused why `page.set_default_timeout()` is not working for `expect()`

Tests should be pretty robust now. Ready to merge.

Closes #768

File: e2e/README.md [hunk 5]
```diff
@@ -16,14 +16,15 @@ All commands/scripts in this README are executed from this folder (/e2e).
 ## Environment Variables
 
 All environment variables are optional, since they have defaults.
+For type details and defaults see [conftest.py](conftest.py)
 
-- `E2E_URL` (Default: localhost:5173)
```
Comment (markus2330) [inline e2e/README.md:20]: Why did you remove the documentation of the defaults? They still seem to exist in e2e/pages/constants.py?

File: e2e/pages/constants.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import os
+
+# in ms, here 30 seconds
```
Comment (markus2330) [inline e2e/pages/constants.py:3]: should be in documentation

**Review by markus2330 — APPROVED** [review]
Great job! :heart:

File: e2e/pages/constants.py (hunk 6 above)
Comment (4ydan) [reply e2e/pages/constants.py:3]: Ups, my bad, somehow lost it between rebasing my commits.

Comment (4ydan) [reply e2e/pages/constants.py:3]: Ah no, I just thought hinting like this `For type details and defaults see [conftest.py](conftest.py)` Is better than duplicating the values.
